File: app/src/main/java/tomaszkopacz/meetbam/activities/MainActivity.java
package tomaszkopacz.meetbam.activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.Menu;
import android.view.MenuItem;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import tomaszkopacz.meetbam.R;
import tomaszkopacz.meetbam.presenters.MainActivityPresenter;
import tomaszkopacz.meetbam.service.PostAdapter;
import tomaszkopacz.meetbam.service.WebService;

public class MainActivity extends AppCompatActivity {

    private MainActivityPresenter presenter;

    @Inject
    WebService service;

    @BindView(R.id.toolbar_main)
    Toolbar toolbar;

    @BindView(R.id.postsRecView)
    RecyclerView postsRecView;

    @BindView(R.id.fab_main)
    FloatingActionButton fab;

    public static final int CAMERA_CODE = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //bind views
        ButterKnife.bind(this);

        //action bar
        setSupportActionBar(toolbar);

        //get dependencies
        ((MainApp)getApplication()).getWebServiceComponent().inject(this);

        //set up presenter
        presenter = new MainActivityPresenter(this, service);

        //prepare recycler view and download posts
        prepareRecView();
        presenter.downloadPostsListToActivity();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.action_user:
                Intent user_intent = new Intent(this, AccountActivity.class);
                startActivity(user_intent);
                break;

            case R.id.action_stats:
                Intent stats_intent = new Intent(this, StatsActivity.class);
                startActivity(stats_intent);
                break;

            case R.id.action_map:
                Intent map_intent = new Intent(this, MapActivity.class);
                startActivity(map_intent);
                break;

            case R.id.action_search:
                Intent search_intent = new Intent(this, SearchActivity.class);
                startActivity(search_intent);
                break;

            case R.id.action_logout:
                presenter.logout();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @OnClick(R.id.fab_main)
    public void onFabClick(View view){
        presenter.photoAttempt();
    }

    /**
     * Configures recycler view. Recycler has fixed size, two spans oriented vertically
     * and default item animator.
     */
    private void prepareRecView(){
        postsRecView.setHasFixedSize(true);
        postsRecView.setLayoutManager(new StaggeredGridLayoutManager(2,
                StaggeredGridLayoutManager.VERTICAL));
        postsRecView.setItemAnimator(new DefaultItemAnimator());
    }

    /**
     * Puts posts stored in adapter into recycler view list.
     * @param adapter
     */
    public void putPosts(PostAdapter adapter){
        postsRecView.setAdapter(adapter);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        // when picture is taken
        if (requestCode == CAMERA_CODE && resultCode == RESULT_OK){
            presenter.afterPhotoTaken();
        }
    }

    /**
     * @return recycler view for post items
     */
    public RecyclerView getPostsRecView(){
        return postsRecView;
    }
}

File: app/src/main/java/tomaszkopacz/meetbam/view/MainActivity.kt
package tomaszkopacz.meetbam.view

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentStatePagerAdapter
import android.support.v4.view.ViewPager
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import kotlinx.android.synthetic.main.activity_main.*
import tomaszkopacz.meetbam.R
import tomaszkopacz.meetbam.presenter.MainActivityPresenter

class MainActivity : AppCompatActivity() {

    private lateinit var presenter: MainActivityPresenter
    private lateinit var pagerAdapter: ScreenSliderPagerAdapter

    companion object {
        const val CAMERA_CODE = 1
        private val NUM_PAGES = 2
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar_main)

        presenter = MainActivityPresenter(this)
        pagerAdapter = ScreenSliderPagerAdapter(supportFragmentManager)
        preparePagerView()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        when (id) {
            R.id.action_user -> {
                val userIntent = Intent(this, AccountActivity::class.java)
                startActivity(userIntent)
            }

            R.id.action_stats -> {
                val statsIntent = Intent(this, StatsActivity::class.java)
                startActivity(statsIntent)
            }

            R.id.action_map -> {
                val mapIntent = Intent(this, MapActivity::class.java)
                startActivity(mapIntent)
            }

            R.id.action_search -> {
                val searchIntent = Intent(this, SearchActivity::class.java)
                startActivity(searchIntent)
            }

            R.id.action_logout -> presenter.logout()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == CAMERA_CODE && resultCode == Activity.RESULT_OK) {
            presenter.photoTaken()
        }
    }

    /*==============================================================================================
                                        PAGER VIEW
    ==============================================================================================*/

    private class ScreenSliderPagerAdapter(fm: FragmentManager): FragmentStatePagerAdapter(fm) {

        override fun getItem(position: Int): Fragment {
            when (position){
                0 -> return MainPhotoFragment()
                1 -> return MainPostsFragment()
            }

            return Fragment()
        }

        override fun getCount(): Int {
            return NUM_PAGES
        }

    }

    private fun preparePagerView(){
        main_viewpager.adapter = pagerAdapter
        main_viewpager.currentItem = 1
        main_viewpager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener{
            override fun onPageScrollStateChanged(state: Int) {
            }

            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
            }

            override fun onPageSelected(position: Int) {
                when(position){
                    0 -> supportActionBar!!.hide()
                    1 -> supportActionBar!!.show()
                }
            }
        })
    }

}
